x/distribution/keeper/allocation.go

```diff
@@ -44,16 +42,16 @@ func (k Keeper) AllocateTokens(ctx sdk.Context, sumPrecommitPower, totalPower in
 	voteMultiplier := sdk.OneDec().Sub(proposerMultiplier).Sub(communityTax)
 
 	// allocate tokens proportionally to voting power
-	// TODO consider parallelizing later, ref https://github.com/cosmos/cosmos-sdk/pull/3099#discussion_r246276376
+	// consider parallelizing later, ref https://github.com/cosmos/cosmos-sdk/pull/3099#discussion_r246276376
 	for _, vote := range votes {
-		validator := k.stakingKeeper.ValidatorByConsAddr(ctx, vote.Validator.Address)
-
-		// TODO likely we should only reward validators who actually signed the block.
-		// ref https://github.com/cosmos/cosmos-sdk/issues/2525#issuecomment-430838701
-		powerFraction := sdk.NewDec(vote.Validator.Power).Quo(sdk.NewDec(totalPower))
-		reward := feesCollected.MulDec(voteMultiplier).MulDec(powerFraction)
-		k.AllocateTokensToValidator(ctx, validator, reward)
-		remaining = remaining.Minus(reward)
+		if vote.SignedLastBlock {
+			powerFraction := sdk.NewDec(vote.Validator.Power).Quo(sdk.NewDec(sumPrecommitPower))
+			reward := feesCollected.MulDec(voteMultiplier).MulDec(powerFraction)
+
+			validator := k.stakingKeeper.ValidatorByConsAddr(ctx, vote.Validator.Address)
+			k.AllocateTokensToValidator(ctx, validator, reward)
```

Comment:
is this call expensive? perhaps we can keep track of total rewards and do a `AllocateTokensToValidator` at the end.

Reply:
Most calls will be to unique validators - but yes, we could do that for a bit of an efficiency gain.